game-core/src/main/java/games/strategy/triplea/ai/pro/util/ProUtils.java

```diff
@@ -218,18 +219,46 @@ public static int getClosestEnemyLandTerritoryDistanceOverWater(final GameData d
   public static boolean isFfa(final GameData data, final PlayerID player) {
     final RelationshipTracker relationshipTracker = data.getRelationshipTracker();
     final Set<PlayerID> enemies = relationshipTracker.getEnemies(player);
-    for (final PlayerID enemy : enemies) {
-      if (relationshipTracker.isAtWarWithAnyOfThesePlayers(enemy, enemies)) {
+    final Set<PlayerID> enemiesWithoutNeutrals =
+        enemies.stream().filter(p -> !isNeutralPlayer(p)).collect(Collectors.toSet());
+    for (final PlayerID enemy : enemiesWithoutNeutrals) {
+      if (relationshipTracker.isAtWarWithAnyOfThesePlayers(enemy, enemiesWithoutNeutrals)) {
         return true;
       }
     }
     return false;
   }
 
+  public static boolean isNeutralLand(final Territory t) {
+    return !t.isWater() && ProUtils.isNeutralPlayer(t.getOwner());
+  }
+
+  /**
+   * Determines whether a player is neutral by checking if all players in its alliance can be considered
+   * neutral as defined by: isPassiveNeutralPlayer OR isHidden and defaultType is AI or DoesNothing.
+   */
   public static boolean isNeutralPlayer(final PlayerID player) {
-    final GameData data = ProData.getData();
-    for (final GameStep gameStep : data.getSequence()) {
-      if (player.equals(gameStep.getPlayerId())) {
+    final GameData data = player.getData();
+    final Set<PlayerID> players = data.getRelationshipTracker().getAllies(player, true);
+    for (final PlayerID p : players) {
+      if (!isPassiveNeutralPlayer(player)
+          && (!p.isHidden() || !(p.isDefaultTypeAi() || p.isDefaultTypeDoesNothing()))) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  /**
+   * Returns true if the player doesn't have a combat move phase.
+   */
+  public static boolean isPassiveNeutralPlayer(final PlayerID player) {
+    if (player.isNull()) {
+      return true;
+    }
+    for (final GameStep gameStep : player.getData().getSequence()) {
+      if (player.equals(gameStep.getPlayerId()) && gameStep.getName().endsWith("CombatMove")
+          && !gameStep.getName().endsWith("NonCombatMove")) {
         return false;
       }
     }
```

Comment:
Here again, a stream use case.

Reply:
Refactored to use streams though needed a little extra magic since this is actually an iterator.